backends/oauth2.go

```diff
@@ -0,0 +1,361 @@
+}
+
+func NewOauth2(authOpts map[string]string, logLevel log.Level) (*Oauth2, error) {
+	log.SetLevel(logLevel)
+
+	var oauth2 = &Oauth2{}
+	oauth2.version = "1.0.0"
+
+	missingOpts := ""
+	oauth2Ok := true
+
+	tokenUrl, ok := authOpts["oauth_token_url"]
+	if ok {
+		oauth2.tokenUrl = tokenUrl
+	} else {
+		oauth2Ok = false
+		missingOpts += " oauth_token_url"
+	}
+
+	if userInfoURL, ok := authOpts["oauth_userinfo_url"]; ok {
+		oauth2.userInfoURL = userInfoURL
+	} else {
+		oauth2Ok = false
+		missingOpts += " oauth_userinfo_url"
+	}
+
+	if cacheDurationSeconds, ok := authOpts["oauth_cache_duration"]; ok {
+		if durationInt, err := strconv.Atoi(cacheDurationSeconds); err == nil {
+			oauth2.cacheDuration = time.Duration(durationInt) * time.Second
+		} else {
+			log.Errorf("unable to parse cacheDurationSeconds: %s", err)
+		}
+	} else {
+		oauth2Ok = false
+		missingOpts += " oauth_cache_duration"
+	}
+
+	if scopes, ok := authOpts["oauth_scopes"]; ok {
+		oauth2.scopesSplit = strings.Split(strings.Replace(scopes, " ", "", -1), ",")
+	} else {
+		log.Infof("no scopes where specified")
+		oauth2.scopesSplit = []string{}
+	}
+
+	oauth2.userCache = make(map[string]userState)
+
+	if !oauth2Ok {
+		return oauth2, errors.Errorf("Oauth2 backend error: missing remote options: %s", missingOpts)
+	}
+
+	return oauth2, nil
+}
+
+func (o *Oauth2) GetUser(username, password, clientid string) (bool, error) {
+	if password == "oauthbearer_empty_password" {
+		return o.createUserWithToken(username, clientid)
+	} else {
+		return o.createUserWithClientCredentials(username, password, clientid)
+	}
+}
+
+func (o *Oauth2) GetSuperuser(username string) (bool, error) {
+	// Function that checks if the user has admin privilies
+	log.Debugf("Checking if user %s is a superuser.", username)
+
+	cache, ok := o.userCache[username]
+	if !ok {
+		log.Infof("no entry in user cache for user %s", username)
+		return false, nil
+	}
+
+	err := o.updateCache(&cache)
+	if err != nil {
+		return false, err
+	}
+
+	log.Debugf("Check for superuser was %t", cache.superuser)
+	o.userCache[username] = cache
+	return cache.superuser, nil
+}
+
+func (o *Oauth2) CheckAcl(username, topic, clientid string, acc int32) (bool, error) {
+	// Function that checks if the user has the right to access a topic
+	log.Debugf("Checking if user %s is allowed to access topic %s with access %d.", username, topic, acc)
+
+	cache, ok := o.userCache[username]
+	if !ok {
+		log.Infof("no entry in user cache for user %s", username)
+		return false, nil
+	}
+
+	err := o.updateCache(&cache)
+	if err != nil {
+		return false, err
+	}
+
+	res := o.checkAccessToTopic(topic, acc, &cache, username, clientid)
+	log.Debugf("ACL check was %t", res)
+	return res, nil
+}
+
+func (o *Oauth2) GetName() string {
+	return "OAuth2 backend, version " + o.version
+}
+
+func (o *Oauth2) Halt() {
+	// Do whatever cleanup is needed.
+}
+
+func (o *Oauth2) updateCache(cache *userState) error {
+
+	if o.cacheIsValid(cache) {
```

Comment:
Mm... I'm not a fan of an `if/else` chain just to log something that could otherwise simply be logged in some kind of `cache valid: %s, o.cacheisValid(cache)` way.
Also, the following multiple calls to `log.Debugf` could just be formatted to be a single one with new lines in the format string.